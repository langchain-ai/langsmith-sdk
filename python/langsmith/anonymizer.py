import copy  # noqa
import re
from abc import abstractmethod
from collections import defaultdict
from typing import Any, Callable, List, Optional, Tuple, TypedDict, Union


class _ExtractOptions(TypedDict):
    maxDepth: Optional[int]
    """
    Maximum depth to traverse to to extract string nodes
    """


class StringNode(TypedDict):
    """String node extracted from the data."""

    value: str
    """String value."""

    path: list[Union[str, int]]
    """Path to the string node in the data."""


def _extract_string_nodes(data: Any, options: _ExtractOptions) -> List[StringNode]:
    max_depth = options.get("maxDepth") or 10

    queue: List[Tuple[Any, int, list[Union[str, int]]]] = [(data, 0, list())]
    result: List[StringNode] = []

    while queue:
        task = queue.pop(0)
        if task is None:
            continue
        value, depth, path = task

        if isinstance(value, dict) or isinstance(value, defaultdict):
            if depth >= max_depth:
                continue
            for key, nested_value in value.items():
                queue.append((nested_value, depth + 1, path + [key]))
        elif isinstance(value, list):
            if depth >= max_depth:
                continue
            for i, item in enumerate(value):
                queue.append((item, depth + 1, path + [i]))
        elif isinstance(value, str):
            result.append(StringNode(value=value, path=path))

    return result


class StringNodeProcessor:
    """Processes a list of string nodes for masking."""

    @abstractmethod
    def mask_nodes(self, nodes: List[StringNode]) -> List[StringNode]:
        """Accept and return a list of string nodes to be masked."""


class ReplacerOptions(TypedDict):
    """Configuration options for replacing sensitive data."""

    maxDepth: Optional[int]
    """Maximum depth to traverse to to extract string nodes."""

    deepClone: Optional[bool]
    """Deep clone the data before replacing."""


class StringNodeRule(TypedDict):
    """Declarative rule used for replacing sensitive data."""

    pattern: re.Pattern
    """Regex pattern to match."""

    replace: Optional[str]
    """Replacement value. Defaults to `[redacted]` if not specified."""


class RuleNodeProcessor(StringNodeProcessor):
    """String node processor that uses a list of rules to replace sensitive data."""

    rules: List[StringNodeRule]

    def __init__(self, rules: List[StringNodeRule]):
        """Initialize the processor with a list of rules."""
        self.rules = rules

    def mask_nodes(self, nodes: List[StringNode]) -> List[StringNode]:
        """Mask nodes using the rules."""
        result = []
        for item in nodes:
            new_value = item["value"]
            for rule in self.rules:
                new_value = rule["pattern"].sub(
                    (
                        rule["replace"]
                        if isinstance(rule["replace"], str)
                        else "[redacted]"
                    ),
                    new_value,
                )
            if new_value != item["value"]:
                result.append(StringNode(value=new_value, path=item["path"]))
        return result


class CallableNodeProcessor(StringNodeProcessor):
    """String node processor that uses a callable function to replace sensitive data."""

    func: Callable[[str, list[Union[str, int]]], str]

    def __init__(self, func: Callable[[str, list[Union[str, int]]], str]):
        """Initialize the processor with a callable function."""
        self.func = func

    def mask_nodes(self, nodes: List[StringNode]) -> List[StringNode]:
        """Mask nodes using the callable function."""
        retval: list[StringNode] = []
        for node in nodes:
            candidate = self.func(node["value"], node["path"])
            if candidate != node["value"]:
                retval.append(StringNode(value=candidate, path=node["path"]))
        return retval


ReplacerType = Union[
    Callable[[str, list[Union[str, int]]], str],
    List[StringNodeRule],
    StringNodeProcessor,
]


def _get_node_processor(replacer: ReplacerType) -> StringNodeProcessor:
    if isinstance(replacer, list):
        return RuleNodeProcessor(rules=replacer)
    elif callable(replacer):
        return CallableNodeProcessor(func=replacer)
    else:
        return replacer


def replace_sensitive_data(
    data: Any, replacer: ReplacerType, options: Optional[ReplacerOptions] = None
) -> Any:
    """Replace sensitive data."""
    nodes = _extract_string_nodes(
        data, {"maxDepth": (options.get("maxDepth") if options else None) or 10}
    )
    mutate_value = copy.deepcopy(data) if options and options["deepClone"] else data

    to_update = _get_node_processor(replacer).mask_nodes(nodes)
    for node in to_update:
        if not node["path"]:
            mutate_value = node["value"]
        else:
            temp = mutate_value
            for part in node["path"][:-1]:
                temp = temp[part]

            last_part = node["path"][-1]
            temp[last_part] = node["value"]

    return mutate_value
