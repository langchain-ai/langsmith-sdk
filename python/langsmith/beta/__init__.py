"""Beta functionality prone to change."""

from langsmith.beta._evals import convert_runs_to_test

__all__ = ["convert_runs_to_test"]
