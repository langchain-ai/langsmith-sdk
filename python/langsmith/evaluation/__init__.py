"""Evaluation Helpers."""

from langsmith.evaluation.evaluator import EvaluationResult, RunEvaluator, run_evaluator
from langsmith.evaluation.string_evaluator import StringEvaluator

__all__ = ["run_evaluator", "EvaluationResult", "RunEvaluator", "StringEvaluator"]
