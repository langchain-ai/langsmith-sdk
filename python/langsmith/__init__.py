"""LangSmith Client."""
from importlib import metadata

from langsmith.client import Client
from langsmith.evaluation.evaluator import EvaluationResult, RunEvaluator
from langsmith.run_trees import RunTree

try:
    __version__ = metadata.version(__package__)
except metadata.PackageNotFoundError:
    # Case where package metadata is not available.
    __version__ = ""

__all__ = [
    "Client",
    "RunTree",
    "__version__",
    "EvaluationResult",
    "RunEvaluator",
]
