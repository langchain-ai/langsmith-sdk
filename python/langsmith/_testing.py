from __future__ import annotations

import atexit
import datetime
import functools
import inspect
import logging
import os
import threading
import uuid
import weakref
from typing import Any, Callable, Optional, Sequence, Tuple, TypeVar, overload

from typing_extensions import TypedDict

from langsmith import client as ls_client
from langsmith import env as ls_env
from langsmith import run_helpers as rh
from langsmith import schemas as ls_schemas
from langsmith import utils as ls_utils

logger = logging.getLogger(__name__)


def _get_experiment_name() -> str:
    # TODO Make more easily configurable
    prefix = ls_utils.get_tracer_project(False) or "TestSuite"
    name = f"{prefix}:{uuid.uuid4().hex[:8]}"
    return name


def _get_test_suite_name() -> str:
    # TODO: This naming stuff is inelegant
    test_suite_name = os.environ.get("LANGCHAIN_TEST_SUITE")
    if test_suite_name:
        return test_suite_name
    if __package__:
        return __package__
    git_info = ls_env.get_git_info()
    if git_info:
        if git_info["remote_url"]:
            repo_name = git_info["remote_url"].split("/")[-1].split(".")[0]
            if repo_name:
                return repo_name
    raise ValueError("Please set the LANGCHAIN_TEST_SUITE environment variable.")


def _get_test_suite(client: ls_client.Client) -> ls_schemas.Dataset:
    test_suite_name = _get_test_suite_name()

    if client.has_dataset(dataset_name=test_suite_name):
        return client.read_dataset(dataset_name=test_suite_name)
    else:
        return client.create_dataset(dataset_name=test_suite_name)


def _start_experiment(
    client: ls_client.Client,
    test_suite: ls_schemas.Dataset,
) -> ls_schemas.TracerSessionResult:
    experiment_name = _get_experiment_name()
    return client.create_project(experiment_name, reference_dataset_id=test_suite.id)


def _get_id(func: Callable) -> uuid.UUID:
    return uuid.uuid5(uuid.NAMESPACE_DNS, f"{func.__module__}.{func.__name__}")


def _end_tests(
    client: ls_client.Client,
    experiment: ls_schemas.TracerSessionResult,
    get_version: Callable[[], Optional[datetime.datetime]],
):
    return client.update_project(
        experiment.id,
        end_time=datetime.datetime.now(datetime.timezone.utc),
        metadata={"dataset_version": get_version()},
    )


class _LangSmithTestSuite:
    _instance = None
    _lock = threading.RLock()

    def __init__(
        self,
        client: Optional[ls_client.Client],
        experiment: ls_schemas.TracerSessionResult,
        dataset: ls_schemas.Dataset,
    ):
        self._client = client or ls_client.Client()
        self._experiment = experiment
        self._dataset = dataset
        self._version: Optional[datetime.datetime] = None
        weakref.finalize(
            self, _end_tests, self._client, self._experiment, self.get_version
        )
        atexit.register(_end_tests, self._client, self._experiment, self.get_version)

    @property
    def id(self):
        return self._dataset.id

    @classmethod
    def get_singleton(cls, client: Optional[ls_client.Client]):
        client = client or ls_client.Client()
        with cls._lock:
            if not cls._instance:
                test_suite = _get_test_suite(client)
                experiment = _start_experiment(client, test_suite)
                cls._instance = cls(client, experiment, test_suite)
        return cls._instance

    @property
    def name(self):
        return self._experiment.name

    def update_version(self, version: datetime.datetime) -> None:
        if self._version is None or version > self._version:
            self._version = version

    def get_version(self) -> Optional[datetime.datetime]:
        return self._version


T = TypeVar("T")
U = TypeVar("U")


class _UTExtra(TypedDict, total=False):
    client: Optional[ls_client.Client]
    id: Optional[uuid.UUID]
    output_keys: Optional[Sequence[str]]
    test_suite_name: Optional[str]


def _ensure_example(
    func: Callable, *args: Any, langtest_extra: _UTExtra, **kwargs: Any
) -> Tuple[ls_schemas.TracerSession, ls_schemas.Example]:
    # 1. check if the id exists.
    # TODOs: Local cache + prefer a peek operation
    client = langtest_extra["client"] or ls_client.Client()
    example_id = langtest_extra["id"] or _get_id(func)
    output_keys = langtest_extra["output_keys"]
    signature = inspect.signature(func)
    # TODO: Support multiple test suites
    test_suite = _LangSmithTestSuite.get_singleton(client)
    try:
        example = client.read_example(example_id=example_id)
    except ls_utils.LangSmithNotFoundError:
        # TODO: support
        # 2. Create the example
        inputs = rh._get_inputs_safe(signature, args, kwargs)
        outputs = {}
        if output_keys:
            for k in output_keys:
                outputs[k] = inputs.pop(k, None)
        example = client.create_example(
            example_id=example_id,
            inputs=inputs,
            # TODO: Handle output_keys
            outputs=outputs,
            dataset_id=test_suite.id,
        )
    test_suite.update_version(example.modified_at)
    return test_suite, example


def _run_test(func, *test_args, langtest_extra: _UTExtra, **test_kwargs):
    test_suite, example = _ensure_example(
        func, *test_args, **test_kwargs, langtest_extra=langtest_extra
    )
    run_id = uuid.uuid4()

    try:
        func_ = func if rh.is_traceable_function(func) else rh.traceable(func)
        func_(
            *test_args,
            **test_kwargs,
            langsmith_extra={
                "run_id": run_id,
                "reference_example_id": example.id,
                "project_name": test_suite.name,
            },
        )
    except BaseException as e:
        client = test_kwargs.get("client") or ls_client.Client()
        client.create_feedback(
            run_id,
            key="pass",
            score=0,
            comment=f"Error: {repr(e)}",
        )
        raise e
    try:
        client = test_kwargs.get("client") or ls_client.Client()
        client.create_feedback(
            run_id,
            key="pass",
            score=1,
        )
    except BaseException as e:
        logger.warning(f"Failed to create feedback for run_id {run_id}: {e}")


@overload
def test_case(
    func: Callable,
) -> Callable: ...


@overload
def test_case(
    *,
    id: uuid.UUID,
    output_keys: Optional[Sequence[str]] = None,
    client: Optional[ls_client.Client] = None,
    # TODO: naming should be consistent probably
    # It's a dataset in the background
    test_suite_name: Optional[str] = None,
) -> Callable[[Callable], Callable]: ...


def test_case(*args, **kwargs):
    """Decorator for defining a LangSmith test case.

    This decorator is used to mark a function as a test case for LangSmith. It ensures
    that the necessary example data is created and associated with the test function.
    The decorated function will be executed as a test case, and the results will be
    recorded and reported by LangSmith.

    Args:
        *args: Positional arguments.
            - If a single callable is provided as a positional argument, it will be
              treated as the test function to be decorated.
        **kwargs: Keyword arguments.
            - id (Optional[uuid.UUID]): A unique identifier for the test case. If not
              provided, an ID will be generated based on the test function's module
              and name.
            - output_keys (Optional[Sequence[str]]): A list of keys to be considered as
              the output keys for the test case. These keys will be extracted from the
              test function's inputs and stored as the expected outputs.
            - client (Optional[ls_client.Client]): An instance of the LangSmith client
              to be used for communication with the LangSmith service. If not provided,
              a default client will be used.
            - test_suite_name (Optional[str]): The name of the test suite to which the
              test case belongs. If not provided, the test suite name will be determined
              based on the environment or the package name.

    Returns:
        Callable: The decorated test function.

    Example:
        To define a test case using the decorator:

        >>> @test_case
        ... def test_addition(a: int, b: int, expected: int):
        ...     assert a + b == expected

        To define a test case with a specific ID and output keys:

        >>> @test_case(id=uuid.uuid4(), output_keys=["result"])
        ... def test_multiplication(a: int, b: int, result: int):
        ...     assert a * b == result

        Using the decorator with pytest fixtures and OpenAI:

        >>> import openai
        >>> import pytest
        >>> from langsmith import test_case
        >>> from langsmith.wrappers import wrap_openai
        >>>
        >>> @test_case
        ... def test_example():
        ...     assert True
        >>>
        >>> @pytest.fixture
        ... def some_input():
        ...     return "Some input"
        >>>
        >>> @test_case
        ... def test_other_example(some_input: str):
        ...     assert "input" in some_input
        >>>
        >>> @test_case
        ... def test_openai_says_hello():
        ...     # Traced code will be included in the test case
        ...     oai_client = wrap_openai(openai.Client())
        ...     response = oai_client.chat.completions.create(
        ...         model="gpt-3.5-turbo",
        ...         messages=[
        ...             {"role": "system", "content": "You are a helpful assistant."},
        ...             {"role": "user", "content": "Say hello!"},
        ...         ],
        ...     )
        ...     assert "hello" in response.choices[0].message.content.lower()


        Run in pytest, or directly run the test functions.

        >>> test_example()
        >>> test_other_example("Some input")
        >>> test_openai_says_hello()
    """
    langtest_extra = _UTExtra(
        id=kwargs.get("id"),
        output_keys=kwargs.get("output_keys "),
        client=kwargs.get("client"),
        test_suite_name=kwargs.get("test_suite_name"),
    )
    if args and callable(args[0]):
        func = args[0]

        @functools.wraps(func)
        def wrapper(*test_args, **test_kwargs):
            _run_test(
                func,
                *test_args,
                **test_kwargs,
                langtest_extra=langtest_extra,
            )

        return wrapper

    def decorator(func):
        @functools.wraps(func)
        def wrapper(*test_args, **test_kwargs):
            _run_test(func, *test_args, **test_kwargs, langtest_extra=langtest_extra)

        return wrapper

    return decorator
