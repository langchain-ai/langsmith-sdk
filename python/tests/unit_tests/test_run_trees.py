from concurrent.futures import ThreadPoolExecutor
from datetime import datetime
from unittest.mock import MagicMock
from uuid import UUID

import pytest

from langsmith import run_trees
from langsmith.client import Client


def test_run_tree_accepts_tpe() -> None:
    mock_client = MagicMock(spec=Client)
    run_trees.RunTree(
        name="My Chat Bot",
        inputs={"text": "Summarize this morning's meetings."},
        client=mock_client,
        executor=ThreadPoolExecutor(),
    )


@pytest.mark.parametrize(
    "inputs, expected",
    [
        (
            "20240412T202937370454Z152ce25c-064e-4742-bf36-8bb0389f8805.20240412T202937627763Zfe8b541f-e75a-4ee6-b92d-732710897194.20240412T202937708023Z625b30ed-2fbb-4387-81b1-cb5d6221e5b4.20240412T202937775748Z448dc09f-ad54-4475-b3a4-fa43018ca621.20240412T202937981350Z4cd59ea4-491e-4ed9-923f-48cd93e03755.20240412T202938078862Zcd168cf7-ee72-48c2-8ec0-50ab09821973.20240412T202938152278Z32481c1a-b83c-4b53-a52e-1ea893ffba51",
            [
                (
                    datetime(2024, 4, 12, 20, 29, 37, 370454),
                    UUID("152ce25c-064e-4742-bf36-8bb0389f8805"),
                ),
                (
                    datetime(2024, 4, 12, 20, 29, 37, 627763),
                    UUID("fe8b541f-e75a-4ee6-b92d-732710897194"),
                ),
                (
                    datetime(2024, 4, 12, 20, 29, 37, 708023),
                    UUID("625b30ed-2fbb-4387-81b1-cb5d6221e5b4"),
                ),
                (
                    datetime(2024, 4, 12, 20, 29, 37, 775748),
                    UUID("448dc09f-ad54-4475-b3a4-fa43018ca621"),
                ),
                (
                    datetime(2024, 4, 12, 20, 29, 37, 981350),
                    UUID("4cd59ea4-491e-4ed9-923f-48cd93e03755"),
                ),
                (
                    datetime(2024, 4, 12, 20, 29, 38, 78862),
                    UUID("cd168cf7-ee72-48c2-8ec0-50ab09821973"),
                ),
                (
                    datetime(2024, 4, 12, 20, 29, 38, 152278),
                    UUID("32481c1a-b83c-4b53-a52e-1ea893ffba51"),
                ),
            ],
        ),
    ],
)
def test_parse_dotted_order(inputs, expected):
    assert run_trees._parse_dotted_order(inputs) == expected


def test_run_tree_events_not_null():
    mock_client = MagicMock(spec=Client)
    run_tree = run_trees.RunTree(
        name="My Chat Bot",
        inputs={"text": "Summarize this morning's meetings."},
        client=mock_client,
        executor=ThreadPoolExecutor(),
        events=None,
    )
    assert run_tree.events == []
