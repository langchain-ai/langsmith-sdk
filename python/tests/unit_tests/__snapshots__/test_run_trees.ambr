# serializer version: 1
# name: test_run_tree_to_dict
  dict({
    'child_runs': list([
      dict({
        'child_runs': list([
        ]),
        'dotted_order': '20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2.20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e3',
        'end_time': None,
        'error': None,
        'events': list([
        ]),
        'extra': dict({
        }),
        'id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e3'),
        'inputs': dict({
        }),
        'name': 'Child',
        'outputs': dict({
        }),
        'reference_example_id': None,
        'run_type': 'chain',
        'serialized': dict({
          'name': 'Child',
        }),
        'session_id': None,
        'session_name': 'default',
        'start_time': FakeDatetime(2023, 1, 1, 0, 0),
        'tags': list([
        ]),
        'trace_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
      }),
      dict({
        'child_runs': list([
        ]),
        'dotted_order': '20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2.20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e4',
        'end_time': None,
        'error': None,
        'events': list([
        ]),
        'extra': dict({
        }),
        'id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e4'),
        'inputs': dict({
        }),
        'name': 'Child2',
        'outputs': dict({
          'summary': "I'm the second child.",
        }),
        'reference_example_id': None,
        'run_type': 'chain',
        'serialized': dict({
          'name': 'Child2',
        }),
        'session_id': None,
        'session_name': 'default',
        'start_time': FakeDatetime(2023, 1, 1, 0, 0),
        'tags': list([
        ]),
        'trace_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
      }),
    ]),
    'dotted_order': '20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2',
    'end_time': None,
    'error': None,
    'events': list([
    ]),
    'extra': dict({
    }),
    'id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
    'inputs': dict({
      'text': "Summarize this morning's meetings.",
    }),
    'name': 'My Chat Bot',
    'outputs': dict({
      'summary': 'You have 3 meetings today.',
    }),
    'parent_run_id': None,
    'reference_example_id': None,
    'run_type': 'chain',
    'serialized': dict({
      'name': 'My Chat Bot',
    }),
    'session_id': None,
    'session_name': 'default',
    'start_time': FakeDatetime(2023, 1, 1, 0, 0),
    'tags': list([
    ]),
    'trace_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
  })
# ---
# name: test_run_tree_to_dict.1
  dict({
    'child_runs': list([
      dict({
        'child_runs': list([
        ]),
        'dotted_order': '20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2.20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e3',
        'end_time': None,
        'error': None,
        'events': list([
        ]),
        'extra': dict({
        }),
        'id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e3'),
        'inputs': dict({
        }),
        'name': 'Child',
        'outputs': dict({
        }),
        'reference_example_id': None,
        'run_type': 'chain',
        'serialized': dict({
          'name': 'Child',
        }),
        'session_id': None,
        'session_name': 'default',
        'start_time': FakeDatetime(2023, 1, 1, 0, 0),
        'tags': list([
        ]),
        'trace_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
      }),
      dict({
        'child_runs': list([
        ]),
        'dotted_order': '20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2.20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e4',
        'end_time': None,
        'error': None,
        'events': list([
        ]),
        'extra': dict({
        }),
        'id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e4'),
        'inputs': dict({
        }),
        'name': 'Child2',
        'outputs': dict({
          'summary': "I'm the second child.",
        }),
        'reference_example_id': None,
        'run_type': 'chain',
        'serialized': dict({
          'name': 'Child2',
        }),
        'session_id': None,
        'session_name': 'default',
        'start_time': FakeDatetime(2023, 1, 1, 0, 0),
        'tags': list([
        ]),
        'trace_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
      }),
    ]),
    'dotted_order': '20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2',
    'end_time': None,
    'error': None,
    'events': list([
    ]),
    'extra': dict({
    }),
    'id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
    'inputs': dict({
      'text': "Summarize this morning's meetings.",
    }),
    'name': 'My Chat Bot',
    'parent_run_id': None,
    'reference_example_id': None,
    'run_type': 'chain',
    'serialized': dict({
      'name': 'My Chat Bot',
    }),
    'session_id': None,
    'session_name': 'default',
    'start_time': FakeDatetime(2023, 1, 1, 0, 0),
    'tags': list([
    ]),
    'trace_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
  })
# ---
# name: test_run_tree_to_dict.2
  dict({
    'child_runs': list([
      dict({
        'child_runs': list([
        ]),
        'dotted_order': '20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2.20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e3',
        'end_time': None,
        'error': None,
        'events': list([
        ]),
        'extra': dict({
        }),
        'id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e3'),
        'inputs': dict({
        }),
        'name': 'Child',
        'outputs': dict({
        }),
        'parent_run_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
        'reference_example_id': None,
        'run_type': 'chain',
        'serialized': dict({
          'name': 'Child',
        }),
        'session_id': None,
        'session_name': 'default',
        'start_time': FakeDatetime(2023, 1, 1, 0, 0),
        'tags': list([
        ]),
        'trace_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
      }),
      dict({
        'child_runs': list([
        ]),
        'dotted_order': '20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2.20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e4',
        'end_time': None,
        'error': None,
        'events': list([
        ]),
        'extra': dict({
        }),
        'id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e4'),
        'inputs': dict({
        }),
        'name': 'Child2',
        'parent_run_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
        'reference_example_id': None,
        'run_type': 'chain',
        'serialized': dict({
          'name': 'Child2',
        }),
        'session_id': None,
        'session_name': 'default',
        'start_time': FakeDatetime(2023, 1, 1, 0, 0),
        'tags': list([
        ]),
        'trace_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
      }),
    ]),
    'dotted_order': '20230101T000000000000Ze0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2',
    'end_time': None,
    'error': None,
    'events': list([
    ]),
    'extra': dict({
    }),
    'id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
    'inputs': dict({
      'text': "Summarize this morning's meetings.",
    }),
    'name': 'My Chat Bot',
    'outputs': dict({
      'summary': 'You have 3 meetings today.',
    }),
    'parent_run_id': None,
    'reference_example_id': None,
    'run_type': 'chain',
    'serialized': dict({
      'name': 'My Chat Bot',
    }),
    'session_id': None,
    'session_name': 'default',
    'start_time': FakeDatetime(2023, 1, 1, 0, 0),
    'tags': list([
    ]),
    'trace_id': UUID('e0ae5bb6-c69a-47bd-89c1-c51ac0cfd9e2'),
  })
# ---
